pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -965,15 +961,23 @@ protected void handleConnect(CommandConnect connect) {
                         remoteAddress, originalPrincipal);
                 }
             }
+            if (service.isAuthorizationEnabled()) {
+                validateRoleAndOriginalPrincipal();
+            }
```

Comment:
Fixed by https://github.com/apache/pulsar/pull/19409 because the authentication protocol is serialized to make it easier to ensure we only verify the roles when authentication is completed.